### PR TITLE
Allow non-4-byte-aligned fills in the compiler.

### PR DESCRIPTION
**Pull request by ScottTodd**
This is only lightly tested at the moment, as not many programs we've looked at have had the potential to generate non-4-byte-aligned fills.

### CODE REVIEW

**Comment by ScottTodd** [comment]
`isConstFloatBuildableWith` assert ([logs](https://source.cloud.google.com/results/invocations/4ffc5fd4-671a-4df4-b173-ae835b92db6c/targets/iree%2Fgcp_ubuntu%2Fbazel%2Flinux%2Fx86-swiftshader%2Fcore%2Fpresubmit/log)) looks a legit issue with the use of an f16 constant. I'll need to debug a bit, so I'll get back to that after reviewing the `stream` dialect PR...

File: iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertCommandBufferOps.cpp [hunk 1]
```diff
@@ -13,6 +13,65 @@ namespace mlir {
...
+      // to f32 changes the bit representation.
+      return op.emitOpError()
+             << "unhandled fill buffer type: " << originalPatternType;
+    }
+    callOperands.push_back(pattern);
+    callOperands.push_back(patternLength);
+
+    auto callOp = rewriter.replaceOpWithNewOp<IREE::VM::CallOp>(
+        op, SymbolRefAttr::get(importOp), importType.getResults(),
+        callOperands);
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Conversion/HALToVM/ConvertCommandBufferOps.cpp:65]: something in here isn't handling type conversions, but I'm not sure what / how to track that down:

```
model.tflite:0:0: error: failed to materialize conversion for result #0 of operation 'arith.constant' that remained live after conversion
model.tflite:0:0: note: called from
model.tflite:0:0: note: see current operation: %139 = "arith.constant"() {value = -1 : i8} : () -> i8
model.tflite:0:0: note: see existing live user here: vm.call @hal.command_buffer.fill_buffer(%ref_94, %ref_91, %zero, %c9409, %c-1_i8, %c1_i32_96) : (!vm.ref<!hal.command_buffer>, !vm.ref<!hal.buffer>, i32, i32, i8, i32) -> ()
model.tflite:0:0: error: conversion to vm.module failed
```

any tips?

**Comment by ScottTodd** [comment]
> Here are the steps to test the implementation with [person_detect](https://github.com/google/iree-samples/blob/670c243/tflitehub/person_detect_test.py#L10) model
> 
> 1. Convert the tflite flatbuffer into mlir with `iree-import-tflite`
> 2. Revert [c4e0f3b](https://github.com/google/iree/commit/c4e0f3b01a874e1bb0b3d3b71d54d5b70f3129e6)
> 
> ```
>  .../iree_compiler/install/bin/iree-translate \
> --iree-mlir-to-vm-bytecode-module \
> --iree-hal-target-backends=dylib-llvm-aot \
> --iree-llvm-target-triple=x86_64 \
> --iree-llvm-link-embedded=true \
> --iree-input-type=tosa \
> --iree-llvm-embedded-linker-path=.../iree_compiler/install/bin/lld
> person_detection.mlir -o /tmp/person_detect_dylib_x86_64_fault.vmfb 
> ```
> 
> 
> ```
> .../iree_compiler/install/bin/iree-run-module \
> --driver=dylib \
> --module_file=/tmp/person_detect_dylib_x86_64_fault.vmfb \
> --entry_function="main" \
> --function_input="1x96x96x1xi8" \
> --trace_execution=true
> ```
> 
> The execution would hang indefinitely with the trace like https://gist.github.com/hcindyl/48f838ee5aa146c7444d1a8be253f235 Specifically, `[module.main+00000408] vm.call @hal.command_buffer.fill_buffer(%r10(!hal.command_buffer/0x0xa44ec0), %r8(!hal.buffer/0x0xa2e030), %i10(0), %i8(9409), %i9(4294967295))`

Thanks! This model works now. I'm having trouble running the mobilenet_quant_test at HEAD and with these changes though. Not really sure if that's related to any of the changes in this area though:

```
I D:\dev\projects\iree\iree\tools\utils\vm_util.cc:185] Creating driver and device for 'dylib'...
Exception thrown at 0x00007FFB347D5397 (iree_dylib_00001170_mem_00000000.dll) in iree-run-module.exe: 0xC00000FD: Stack overflow (parameters: 0x0000000000000001, 0x00000067BF103000).
The program '[4464] iree-run-module.exe' has exited with code 0 (0x0).
```

**Comment by hcindyl** [comment]
> --trace_execution=true

Try to remove `--trace_execution=true` as the flag would create a much deeper stack that could cause overflow. However, it will make the runtime debug more difficult though.

**Comment by ScottTodd** [comment]
> > --trace_execution=true
> 
> Try to remove `--trace_execution=true` as the flag would create a much deeper stack that could cause overflow. However, it will make the runtime debug more difficult though.

I'm not using `--trace_execution`. Just `--entry_function=main --function_input="1x224x224x3xi8"` for `func @main(%arg0: tensor<1x224x224x3xui8>` (`ui8` wasn't parsing)

File: iree/test/e2e/models/collatz.mlir [hunk 2]
```diff
@@ -1,4 +1,5 @@
 // RUN: iree-run-mlir --iree-input-type=mhlo -iree-hal-target-backends=vmvx %s | IreeFileCheck %s
+// RUN: [[ $IREE_VULKAN_DISABLE == 1 ]] || (iree-run-mlir --iree-input-type=mhlo %s -iree-hal-target-backends=vulkan-spirv | IreeFileCheck %s)
```
[the remark this answers is not in the document]
Comment (ScottTodd) [reply iree/test/e2e/models/collatz.mlir:2]: CPU path seems to work with the "Don't outline splat constants." change, but Vulkan is still showing numerical issues that I'm tracking down. The person detect model doesn't compile through Vulkan though? :/ `failed to legalize operation 'scf.while'` ([logs here](https://gist.github.com/ScottTodd/dbfd18a84bc839bab0a12ecc71dd9b3a))

[the remark this answers is not in the document]
Comment (hcindyl) [reply iree/test/e2e/models/collatz.mlir:2]: person_detect model is not targeted for the vulkan backend, and we don't have tests for that.

**Comment by hcindyl** [comment]
> > > --trace_execution=true
> > 
> > 
> > Try to remove `--trace_execution=true` as the flag would create a much deeper stack that could cause overflow. However, it will make the runtime debug more difficult though.
> 
> I'm not using `--trace_execution`. Just `--entry_function=main --function_input="1x224x224x3xi8"` for `func @main(%arg0: tensor<1x224x224x3xui8>` (`ui8` wasn't parsing)

I think the function_input format should be `1x224x224x3xu8` based on https://github.com/google/iree/blob/main/iree/hal/string_util.c#L108

Another way to test is to run the test at https://github.com/google/iree-samples/blob/main/tflitehub/mobilenet_quant_test.py with splat fill_buffer enabled in `iree-translate`

**Pull request merged**